### PR TITLE
Emote keybinds

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

You can set keybinds to proc emotes. A 2 second cooldown between two identical emotes.
You can also set 10 custom me/say emotes, and bind them to whatever

Sans titre

## Why It's Good For The Game

Death to macro. Also the cooldown should prevent being muted : https://github.com/tgstation/TerraGov-Marine-Corps/issues/6730

## Changelog
:cl:
add: Emotes keybinds, 10 custom me/say emotes keybinds
/:cl:

### CODE REVIEW

File: code/datums/custom_emotes.dm [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/datum/custom_emote
+	/// An id to be able to distinguish emotes
+	var/id = 0
+	///Message displayed when emote is used
+	var/message = "" 
+	/// Cooldown between two use of that emote. Every emote has its own coodldown
+	var/cooldown = 5 SECONDS
+	/// If this custom emote is a say or a me
+	var/spoke_emote = TRUE
```
Comment (TiviPlus) [inline code/datums/custom_emotes.dm:9]:
```suggestion
	/// If this custom emote is a say or a me
	var/spoken_emote = TRUE
```
a spoke is part of a wheel

File: code/datums/custom_emotes.dm [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	return TRUE
+
+/// Check if that emote was used recently
+/datum/custom_emote/proc/check_cooldown(mob/user)
+	if(user.emotes_used && user.emotes_used[src] + cooldown > world.time)
+		return FALSE
+	if(!user.emotes_used)
+		user.emotes_used = list()
+	user.emotes_used[src] = world.time
+	return TRUE
```
Comment (TiviPlus) [inline code/datums/custom_emotes.dm:41]: couldnt these be easily inlined?
also why that first istype how is a nonmob running emotes?

File: code/modules/mob/mob_defines.dm [hunk 3]
```diff
@@ -91,3 +91,6 @@
 
 	/// Can they interact with station electronics
 	var/has_unlimited_silicon_privilege = 0
+
+	/// Used for tracking last uses of emotes for cooldown purposes
+	var/list/emotes_used
```
Comment (TiviPlus) [inline code/modules/mob/mob_defines.dm:96]: Why isnt this cooldown a timer signal

**Review by TiviPlus — CHANGES_REQUESTED** [review]
@SplinterGP tgui review pls

File: code/datums/custom_emotes.dm [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	var/spoken_emote = TRUE
+
+/// Run the custome emote
+/datum/custom_emote/proc/run_custom_emote(mob/user)
+	if(!message)
+		return
+	if(TIMER_COOLDOWN_CHECK(user, "Custom_emotes[id]"))
+		to_chat(user, "<span class='notice'>You used that emote too recently.</span>")
+		return
+	TIMER_COOLDOWN_START(user, "Custom_emotes[id]", cooldown)
```
Comment (TiviPlus) [inline code/datums/custom_emotes.dm:18]: lowercase these stings

File: code/datums/emotes.dm [hunk 5]
```diff
@@ -85,6 +87,14 @@
 	else
 		user.visible_message(msg, visible_message_flags = EMOTE_MESSAGE, emote_prefix = prefix)
 
+/// For handling emote cooldown, return true to allow the emote to happen
+/datum/emote/proc/check_cooldown(mob/user, intentional)
+	if(!intentional)
+		return TRUE
+	if(TIMER_COOLDOWN_CHECK(user, "Emote [key]"))
+		return FALSE
+	TIMER_COOLDOWN_START(user, "Emote [key]", cooldown)
```
Comment (TiviPlus) [inline code/datums/emotes.dm:96]: ditto

File: code/modules/client/preferences_ui.dm [hunk 6]
```diff
@@ -619,6 +627,25 @@
...
+			emote.message = params["sentence"]
+			custom_emotes[id] = emote
+
+		if("setEmoteType")
+			var/kb_name = params["name"]
+			if(!kb_name)
+				return
+			var/list/part = splittext(kb_name, ":")
+			var/id = text2num(part[2])
+			var/datum/custom_emote/emote = custom_emotes[id]
```
Comment (TiviPlus) [inline code/modules/client/preferences_ui.dm:646]: is there no limit to the length of the emote?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/client/preferences_ui.dm (hunk 6 above)
Comment (BraveMole) [reply code/modules/client/preferences_ui.dm:646]: 300 , it's in the TGUI

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**